DNN Platform/Library/Services/Installer/Installers/CleanupInstaller.cs

```diff
@@ -102,6 +152,47 @@ public override void UnInstall()
         {
         }
 
+        /// <summary>
+        /// Adds a folder path to the list.
+        /// </summary>
+        /// <param name="path">The folder path.</param>
+        internal void ProcessFolder(string path)
+        {
+            this.Folders.Add(path);
+        }
+
+        /// <summary>
+        /// Validates a folder path for cleanup.
+        /// </summary>
+        /// <param name="path">The folder oath to validate.</param>
+        /// <returns>Whether or not the folder path is valid.</returns>
+        internal bool IsValidFolderPath(string path)
+        {
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                return false;
+            }
+
+            string normalized;
+            try
+            {
+                normalized = Path.GetFullPath(path);
+            }
+            catch
+            {
+                return false;
+            }
+
+            var appPath = this.applicationStatusInfo.ApplicationMapPath;
+            var fullPath = Path.Combine(appPath, normalized);
+            if (!normalized.StartsWith(appPath, StringComparison.InvariantCultureIgnoreCase) || fullPath == appPath)

```

Comment:
I have a feeling we need to be more protective with this validation test as it would be very easy to corrupt an installation with this.

Values such as

* /DesktopModules
* /Portals/0
* /Bin

Or otherwise, and since this is a recursive delete, that doesn't expect folders to be empty I think we need to prevent at least our critical directories from being cleaned?

Reply:
Yes, good point, although critical files could still get deleted using the `glob` parameter.
Anyway, I have changed the original indiscriminate delete now with a new method that only deletes _empty_ directories. This avoids the risk of any files being deleted by CleanupInstaller.